armi/reactor/grids/structuredgrid.py

```diff
@@ -0,0 +1,511 @@
+        """Return the coordinates of the center of the mesh cell at the given given indices in cm."""
+        indices = numpy.array(indices)
+        return self._evaluateMesh(
+            indices, self._centroidBySteps, self._centroidByBounds
+        )
+
+    def getCellBase(self, indices) -> numpy.ndarray:
+        """Get the mesh base (lower left) of this mesh cell in cm"""
+        indices = numpy.array(indices)
+        return self._evaluateMesh(
+            indices, self._meshBaseBySteps, self._meshBaseByBounds
+        )
+
+    def getCellTop(self, indices) -> numpy.ndarray:
+        """Get the mesh top (upper right) of this mesh cell in cm"""
+        indices = numpy.array(indices) + 1
+        return self._evaluateMesh(
+            indices, self._meshBaseBySteps, self._meshBaseByBounds
+        )
+
+    def _evaluateMesh(self, indices, stepOperator, boundsOperator) -> numpy.ndarray:
+        """
+        Evaluate some function of indices on this grid.
+
+        Recall from above that steps are mesh centered and bounds are mesh edged.
+
+        Notes
+        -----
+        This method may be able to be simplified. Complications from arbitrary
+        mixtures of bounds-based and step-based meshing caused it to get bad.
+        These were separate subclasses first, but in practice almost all cases have some mix
+        of step-based (hexagons, squares), and bounds based (radial, zeta).
+
+        Improvements welcome!
+        """
+        boundCoords = []
+        for ii, bounds in enumerate(self._bounds):
+            if bounds is not None:
+                boundCoords.append(boundsOperator(indices[ii], bounds))
+
+        # limit step operator to the step dimensions
+        stepCoords = stepOperator(numpy.array(indices)[self._stepDims])
+
+        # now mix/match bounds coords with step coords appropriately.
+        result = numpy.zeros(len(indices))
+        result[self._stepDims] = stepCoords
+        result[self._boundDims] = boundCoords
+        return result + self._offset
+
+    def _centroidBySteps(self, indices):
+        return numpy.dot(self._unitSteps, indices)
+
+    def _meshBaseBySteps(self, indices):
+        return (
+            self._centroidBySteps(indices - 1) + self._centroidBySteps(indices)
+        ) / 2.0
+
+    @staticmethod
+    def _centroidByBounds(index, bounds):
+        if index < 0:
+            # avoid wrap-around
+            raise IndexError("Bounds-defined indices may not be negative.")
+        return (bounds[index + 1] + bounds[index]) / 2.0
+
+    @staticmethod
+    def _meshBaseByBounds(index, bounds):
+        if index < 0:
+            raise IndexError("Bounds-defined indices may not be negative.")
+        return bounds[index]
+
+    @staticmethod
+    def getNeighboringCellIndices(i, j=0, k=0):
+        """Return the indices of the immediate neighbors of a mesh point in the plane."""
+        return ((i + 1, j, k), (1, j + 1, k), (i - 1, j, k), (i, j - 1, k))
```

Comment:
@drewj-usnctech the second element looks wrong. Should be `(i, j + 1, k)`, not `(1, j + 1, k)`, right?

I came across this while writing a unit test for this with a Cartesian grid: 

https://github.com/terrapower/armi/blob/14c951a05214154a1e6bfb43c728aa714084cea7/armi/reactor/reactors.py#L1676-L1678

Reply:
https://github.com/terrapower/armi/pull/1614

Reply:
https://github.com/terrapower/armi/pull/1614#issuecomment-1915149546